src/inputs/TreeSelectField/TreeSelectField.tsx

```diff
@@ -0,0 +1,540 @@
+
+          const { option, parents } = maybeOption;
+
+          // If the option has children, then we need to unselect those children as well
+          if (option.children && option.children.length > 0) {
+            const childrenKeys = option.children.flatMap(flattenOptions).map((o) => valueToKey(getOptionValue(o)));
+            [key, ...childrenKeys].forEach(removedKeys.add, removedKeys);
+          }
+
+          // If the option has parents, then we need to unselect those parents as well
+          if (parents.length > 0) {
+            const parentKeys = parents.map((o) => valueToKey(getOptionValue(o)));
+            [key, ...parentKeys].forEach(removedKeys.add, removedKeys);
+          }
+        });
+
+        // Create the lists to update our TreeState with
+        const selectedKeys = new Set([...existingKeys, ...addedKeys].filter((x) => !removedKeys.has(x)));
+        const selectedOptions = fieldState.allOptions
+          .flatMap(flattenOptions)
+          .filter((o) => selectedKeys.has(valueToKey(getOptionValue(o))));
+
+        // Our `onSelect` callback provides three things:
+        // 1. ALL selected values + options
+        // 2. Only the top level selected values + options
+        // 3. Only the leaf selected values + options
+
+        // For the top level and leaf selections, we need to do some extra work to determine which options are the top level and leaf.
+        function getTopLevelSelections(o: NestedOption<O>): NestedOption<O>[] {
+          // If this element is already selected, return early. Do not bother looking through children.
+          if (selectedKeys.has(valueToKey(getOptionValue(o)))) return [o];
+          // If this element has no children, then look through the children for top level selected options.
+          if (o.children) return [...o.children.flatMap(getTopLevelSelections)];
+          return [];
+        }
+        const rootOptions = fieldState.allOptions.flatMap(getTopLevelSelections);
+        const rootValues = rootOptions.map(getOptionValue);
+
+        // Finally get the list of options that are just the "leaf" options, meaning they have no children.
+        const leafOptions = selectedOptions.filter((o) => !o.children || o.children.length === 0);
+        const leafValues = leafOptions.map(getOptionValue);
+
+        setFieldState((prevState) => ({
+          ...prevState,
+          selectedKeys: [...selectedKeys],
+          selectedOptions,
+        }));
+
+        onSelect({
+          all: { values: [...selectedKeys].map((key) => keyToValue(key)), options: selectedOptions },
+          leaf: { values: leafValues, options: leafOptions },
+          root: { values: rootValues, options: rootOptions },
+        });
+      }
+    },
+  });
+
+  function resetField() {
+    const { inputValue, selectedOptions } = fieldState;
+    if (inputValue !== "" || (selectedOptions.length === 1 && inputValue !== getOptionLabel(selectedOptions[0]))) {
```

Comment:
maybe an an in-source comment here explaining what `reset` actually means? I'm not quite following the logic for this `inputValue` - I naively expect it to just clear on a `reset` type action?